compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -3031,6 +3032,42 @@ declare_lint! {
 
 }
 
+declare_lint! {
+    /// The `disjoint_capture_auto_traits` lint detects when a variable implements a trait, but
+    /// one of its field does not and is captured and used with the assumption that said field
+    /// implements the same trait as the root variable.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(disjoint_capture_auto_traits)]
+    /// use std::thread;
+    ///
+    /// struct Pointer (*mut i32);
+    /// unsafe impl Send for Pointer {}
+    ///
+    /// fn main() {
+    ///     let mut f = 10;
+    ///     let fptr = Pointer(&mut f as *mut i32);
+    ///     thread::spawn(move || unsafe {
+    ///         *fptr.0 = 20;
+    ///     });
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// In the above example `fptr.0` cannot be sent between thread safely as the field
+    /// is captured and detected as not implementing Send if the feature
```

Comment:
I the above example `fptr.0` is captured when feature .... is enabled. The field is of type `*mut i32` which doesn't implement Send, making the code invalid